util/queue/persist.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2022 The TestGrid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+	"github.com/sirupsen/logrus"
+)
+
+// PersistClient contains interfaces for reading from and writing to a Path.
+type PersistClient interface {
+	gcs.Uploader
+	gcs.Opener
+}
+
+// Fixer will adjust the queue until the context expires.
+type Fixer func(context.Context, *Queue) error
+
+// FixPersistent persists the queue to the remote path every tick.
+//
+// The first time it will load the state. Thereafter it will save the state.
+// This includes restarts due to expiring contexts -- it will just load once.
+func FixPersistent(logr logrus.FieldLogger, client PersistClient, path gcs.Path, tick <-chan time.Time) Fixer {
```

Comment:
(a) and (b)

I originally had this as `time.Duration` instead of `<-chan time.TIme` but then this made testing flaky and requiring sleeps. With this pattern we can test immediately and precisely:

```golang

fix := Fix(tc.client) // initial stored state
wg.Add(1)
go func() {
  for _, tick := range tc.ticks {
    ch <- tick
  }
  wg.Done()
}
fix(tc.q) // initial q state + fixing ticks
wg.Wait()
// verify uploaded files
// verify queue state
```


The signature isn't that much more complicated: `FixPersisent(..., time.NewTicker(time.Minute))` instead of `FixPersistent(..., time.Minute)` so seems like a good trade off for the increase in testing control.